lib/p2p/Pool.ts

```diff
@@ -191,44 +218,44 @@ class Pool extends EventEmitter {
       case PacketType.ORDERS: {
         const orders = (packet as OrdersPacket).body!;
         orders.forEach((order) => {
-          this.emit('packet.order', { ...order, peerId: peer.id } as PeerOrder);
+          this.emit('packet.order', { ...order, peerPubKey: peer.nodePubKey } as PeerOrder);
         });
         break;
       }
-      case PacketType.GET_HOSTS: {
-        peer.sendHosts(this.hosts.toHostFactoryArray(), packet.header.id);
+      case PacketType.GET_NODES: {
+        peer.sendNodes(this.nodes.toFactoryArray(), packet.header.id);
         break;
       }
-      case PacketType.HOSTS: {
-        const hosts = (packet as HostsPacket).body!;
-        hosts.forEach(async (host) => {
-          try {
-            await this.addOutbound(new SocketAddress(host.address, host.port));
-          } catch (err) {
-            this.logger.info(err);
+      case PacketType.NODES: {
+        const nodes = (packet as NodesPacket).body!;
+        nodes.forEach(async (node) => {
+          if (!this.nodes.has(node.nodePubKey) && node.nodePubKey !== this.handshakeData.nodePubKey && node.addresses.length > 0) {
+            // attempt to connect to this node if we've never connected before
+            this.connectNode(node);
           }
         });
         break;
       }
     }
   }
 
-  private handleOpen = async (peer: Peer, handshakeState: HandshakeState): Promise<void> => {
-    this.setPeerHost(peer, handshakeState.listenPort);
-
-    // request peer's orders and known hosts
-    peer.sendPacket(new GetOrdersPacket());
-    peer.sendPacket(new GetHostsPacket());
-  }
-
-  private setPeerHost = async (peer: Peer, listenPort?: number): Promise<void> => {
-    if (!peer.inbound) {
-      const host = await this.hosts.getOrCreateHost(peer);
-      peer.setHost(host);
-    } else if (listenPort) {
-      const socketAddress = new SocketAddress(peer.socketAddress.address, listenPort);
-      const host = await this.hosts.getOrCreateHost(peer);
-      peer.setHost(host);
+  private handleOpen = async (peer: Peer): Promise<void> => {
+    if (this.nodes.isBanned(peer.nodePubKey!)) {
+      // TODO: Ban IP address for this session if banned peer attempts repeated connections.
```

Comment:
I think in both places, or rather right after the handshake. The if statement here is supposed to catch peers who are banned who connect to us anyway. The Peer class doesn't know about which nodes are banned which is why the logic is here. But I think if the handshake is bad or invalid, we can also put logic to ban before emitting the `open` event.

Reply:
We were talking earlier about blocking inbound peers prior to handshake, based on IP address. Just wanted to know if we plan to support that, because if we do, we'll need to query nodes by IP address. If we don't, we're more vulnerable to DDoS.

This doesn't apply for banning within session, if we'll just collect banned IP addresses.
But we might want to apply this to blacklist/whitelist modes, which should persist, and might refer to a node entity - not strictly an IP address.

But anyway, we'll deal with that when we'll get to it.